src/openms_gui/source/VISUAL/SpectrumCanvas.cpp

```diff
@@ -391,60 +391,120 @@ namespace OpenMS
 
   bool SpectrumCanvas::addLayer(ExperimentSharedPtrType map, ODExperimentSharedPtrType od_map, const String & filename)
   {
-    layers_.resize(layers_.size() + 1);
-    layers_.back().param = param_;
-    layers_.back().filename = filename;
-    layers_.back().setPeakData(map);
-    layers_.back().setOnDiscPeakData(od_map);
-
-    if (layers_.back().getPeakData()->getChromatograms().size() != 0 
-        && layers_.back().getPeakData()->size() != 0)
+    LayerData new_layer;
+    new_layer.param = param_;
+    new_layer.filename = filename;
+    new_layer.name = QFileInfo(filename.toQString()).baseName();
+    new_layer.setPeakData(map);
+    new_layer.setOnDiscPeakData(od_map);
+
+    if (new_layer.getPeakData()->getChromatograms().size() != 0 
+        && new_layer.getPeakData()->size() != 0)
     {
       // TODO : handle this case better
       OPENMS_LOG_WARN << "Your input data contains chromatograms and spectra, falling back to display spectra only." << std::endl;
     }
 
-    if (layers_.back().getPeakData()->getChromatograms().size() != 0 
-        && layers_.back().getPeakData()->size() == 0)
+    if (new_layer.getPeakData()->getChromatograms().size() != 0 
+        && new_layer.getPeakData()->size() == 0)
     {
-      layers_.back().type = LayerData::DT_CHROMATOGRAM;
+      new_layer.type = LayerData::DT_CHROMATOGRAM;
     }
     else
     {
-      layers_.back().type = LayerData::DT_PEAK;
+      new_layer.type = LayerData::DT_PEAK;
+    }
+
+    // insert after last peak or chomatogram layer
+    for (auto it = layers_.rbegin(); it != layers_.rend(); ++it)
+    {
+      if (it->type == LayerData::DT_PEAK 
+        || it->type == LayerData::DT_CHROMATOGRAM)
+      {
+        layers_.insert(it.base(), new_layer);
+        return finishAdding_();
+      }
     }
+
+    // insert at front
+    layers_.insert(layers_.begin(), new_layer);
     return finishAdding_();
   }
 
   bool SpectrumCanvas::addLayer(FeatureMapSharedPtrType map, const String & filename)
   {
-    layers_.resize(layers_.size() + 1);
-    layers_.back().param = param_;
-    layers_.back().filename = filename;
-    layers_.back().getFeatureMap() = map;
-    layers_.back().type = LayerData::DT_FEATURE;
+    LayerData new_layer;
+    new_layer.param = param_;
+    new_layer.filename = filename;
+    new_layer.name = QFileInfo(filename.toQString()).baseName();
+    new_layer.getFeatureMap() = map;
+    new_layer.type = LayerData::DT_FEATURE;
+
+    // draw features after last feature map (or if not present, after last peak map)
+    for (auto it = layers_.rbegin(); it != layers_.rend(); ++it)
+    {
+      if (it->type == LayerData::DT_PEAK 
+        || it->type == LayerData::DT_FEATURE)
+      {
+        layers_.insert(it.base(), new_layer);
+        return finishAdding_();
+      }
+    }
+    // insert at front
+    layers_.insert(layers_.begin(), new_layer);
```

Comment:
The need to check for all previous layer types seems very brittle and easy to get wrong.
How about we introduce an ordering here. The enum `LayerData::DataType` would be good, you just need to reorder the elements to the order we want and document that the order has meaning.
Then its very short to write (untested code ahead):
```
auto it = std::find_if(layers_.rbegin(), layers_.rend(), [](const LayerData& l) 
      { return l.type <= LayerData::DT_FEATURE; };
layers_.insert(it, std::move(new_layer);
return finishAdding_();
```

Reply:
Yeah I though about that and will probably implement this.